src/backend/tests/unit/test_setup_superuser.py

```diff
@@ -130,3 +134,110 @@ async def test_teardown_superuser_no_default_superuser():
+    assert mock_session.rollback.call_count == 1  # Session was rolled back after IntegrityError
+    assert mock_get_user_by_username.call_count == 2  # Called twice: initial check + after rollback
+
+
+@pytest.mark.asyncio
+async def test_create_super_user_race_condition_no_user_found():
+    """Test that create_super_user re-raises exception if no user is found after IntegrityError."""
+    # Mock the database session
+    mock_session = AsyncMock()
+
+    # Mock get_user_by_username to always return None (even after rollback)
+    mock_get_user_by_username = AsyncMock()
+    mock_get_user_by_username.side_effect = [None, None]  # None for initial check and after rollback
+
+    # Mock other dependencies
+    mock_get_password_hash = MagicMock(return_value="hashed_password")
+    mock_user = MagicMock(spec=User)
+
+    # Set up scenario where IntegrityError occurs but no user is found afterward
+    integrity_error = IntegrityError("statement", "params", "orig")
+    mock_session.commit.side_effect = integrity_error
+
+    with (
+        patch("langflow.services.auth.utils.get_user_by_username", mock_get_user_by_username),
+        patch("langflow.services.auth.utils.get_password_hash", mock_get_password_hash),
+        patch("langflow.services.database.models.user.model.User", return_value=mock_user),
+    ):
+        with pytest.raises(IntegrityError):
+            await create_super_user("testuser", "password", mock_session)
+
+    # Verify rollback was called but exception was re-raised
+    assert mock_session.rollback.call_count == 1
+    assert mock_get_user_by_username.call_count == 2  # Initial + after rollback
+
+
+@pytest.mark.asyncio
+async def test_create_super_user_concurrent_workers():
+    """Test multiple concurrent calls to create_super_user with the same username."""
+    # This would require a real database to properly test, but we can simulate
+    # the behavior with mocks to verify the logic works correctly
+
+    mock_session1 = AsyncMock()
+    mock_session2 = AsyncMock()
+
+    # Create mock users
+    mock_user = MagicMock(spec=User)
+    mock_user.username = "admin"
+    mock_user.is_superuser = True
+
+    mock_get_user_by_username = AsyncMock()
+
+    # Worker 1 succeeds, Worker 2 gets IntegrityError then finds existing user
+    mock_session1.commit.return_value = None  # Success
+    mock_session2.commit.side_effect = IntegrityError("statement", "params", "orig")  # Race condition
+
+    # get_user_by_username returns None initially, then the created user for worker 2
+    mock_get_user_by_username.side_effect = [None, None, mock_user]
+
+    with patch("langflow.services.auth.utils.get_user_by_username", mock_get_user_by_username):
+        # Simulate concurrent execution
+        result1 = await create_super_user("admin", "password", mock_session1)
+        result2 = await create_super_user("admin", "password", mock_session2)
```

Comment:
What if we create both tasks and call `asyncio.gather`? That might be closer to running two workers.